### ISSUE TITLE
Command - Undo Replace All

### ISSUE THREAD

**Issue by lukaszpolowczyk** [Microsoft/vscode#47783]
In Search Panel should be button "Undo Replace All", whitch undo all "replaced" changes in all files.
Should to be work like standard undo function, but only for files whitch "Replace All" made a changes.

It looks something simple to introduce. Am I right?

**Comment by lukaszpolowczyk** [comment]
Bot, as usual, will look for something that has nothing to do with my entry ...

**Comment by spacemansi23** [comment]
#47783

**Comment by spacemansi23** [comment]
Command-undo Replace all

**Comment by sandy081** [comment]
Might not be trivial because there could be edits to the file at the location where replace has happened and undo has to ignore that.

**Comment by lukaszpolowczyk** [comment]
@sandy081 So the button should work only until some manual change is made to these files.
It would be useful anyway.

**Comment by Nowaker** [comment]
IntelliJ IDEA family of IDEs has this feature.

**Comment by RavenHursT** [comment]
Starting to see the above comment as the reality more and more... :-/

**Comment by opensourcegeek** [comment]
It would be really useful to have this feature, I'm `git stash`ing after global replace all at the moment to work around (not ideal).

**Comment by ischmal** [comment]
I support this for instances when I'm not paying attention and accidentally do find-replace to an entire project instead of just the open file. Cleaning up the mess is both humiliating and absurdly tedious.

**Comment by gudlakalyan** [comment]
I hope we cann't undo all files , If this possibility there mean its great . please share the same how to undo all replaced files.

Replace all feature is very helpful, but we should be careful when applying this with all files. 

Better we should apply some filters like file extension , include and/or exclude files , I attached reference screenshot also.

![Visual Code Replace all feature](https://user-images.githubusercontent.com/53933831/62765041-e9226f00-baac-11e9-93ae-c18732245fb9.png)

**Comment by CHN-STUDENT** [comment]
I think perhaps we need this command, because if i replace some thing by mistake,i need to replace them by replacing command again.
Or Anyone who have other way?
Thanks.

**Comment by mortizw** [comment]
At least there should be a warning if you accidentally forget to fill out the lower field and replace every occurrence of the searched item with an empty string.

**Comment by lukaszpolowczyk** [comment]
@mortizw I mean situations when I have already changed something and LATER I realized that I did it incorrectly or unnecessarily.

A warning is not enough for this.

**Comment by Nowaker** [comment]
> Might not be trivial because there could be edits to the file at the location where replace has happened and undo has to ignore that.

@sandy081 Not really. The "replace all" action should be no different from a manual change that lands to the "undo" stack. Undo is a stack and you can only undo the most recent change. "Replace all" action should land on this very stack, and be undoable when it's on the top of the stack. In this case, there's "no [other] edits to ignore". That's how it works in IntelliJ IDEA.

**Comment by raikel** [comment]
> I support this for instances when I'm not paying attention and accidentally do find-replace to an entire project instead of just the open file. Cleaning up the mess is both humiliating and absurdly tedious.

Well, I just accidentally modified 423 lines of code in 123 files.

**Comment by mpfaff** [comment]
And I just screwed up my regex, and now have to fix 132 changes manually.

**Comment by RavenHursT** [comment]
This is why I just use JetBrains IDEs for development now...  #youGetWhatYouPayFor

**Comment by mpfaff** [comment]
@RavenHursT I mean, you're right, but there's no *way* I'm going to pay JetBrains' prices. I'd love to donate a bit to VSCode, but not nearly what JetBrains charge (at least not as a lump sum).

Besides, JetBrains IDEs have many other issues that bother me much more:

- resource hungry
- slow
- take up a large amount of disk space
- dialogs that butcher tiling WMs
- themes that don't work on the whole window

**Comment by hamikm** [comment]
Chiming in — this would be a great (albeit nontrivial) feature

**Comment by vywern** [comment]
Thanking myself for googling this up right before a major replace all...

**Comment by emmanuelnk** [comment]
Chiming in too. 

Of course I could be wrong but I don't see why this is a non-trivial feature to implement. If I do a replace all and I go to an individual file, I can `CTRL+Z` and undo the replace action in that individual file, and thus I can do this in every file that has had the replace action. So it would be as simple as undoing the last function in every file that was replaced. 

Definitely a much needed feature considering how powerful vscode's replace all feature is.

**Comment by MohamedLamineAllal** [comment]
A solution would be to just use the normal undo stack! And to keep a mapping to what a replace all had as entries!

When the replace all undo is triggered! The mapping is used! We check for the whole list of files that were modified! And See if they were top in stack! If all are! And in the right stacked order for all file! Then it's doable and it can be handled!

Now for performance! One can optimize by listening to redo/undo stack of each concerned file!
If no change happened then cool! If some changes happened! Till they got canceled! (we are tracking the hassling block ! That give us the performance point! you don't go and check all! you track it on the go)
One can have even a walk thorugh file by file guiding to undo! (with undo, redo) (so it give the possiblity for copying ! undo! And then when all resolved! the button for undo replace all become enabled! (kind of like a conflict resolution in git walk through)! Simply guided and accessible!

More thinking can go in the thing!

Otherwise too a simple format! Straight forward! Not the most cool thing! But very helpful! Simply make it happen when no changes were made! Because a lot of times one hit replace all! And at the instant it's done! You know you made a mistake! You just want to cancel that! May be that's the most common scenario! (for this only! The implementation is straight forward! After a replace all! All the changes are saved temporally! If any of the file is modified! And changed! Then it becomes impossible and the changes get lost! Something like that!

EVen though using the stack is the cool thing! And require some work! For a cool handling! And simple walk through! In the worst  cases!

Now clearly it's a matter of hands! (Time and will) ! Because that can go too as an extension! Why i will not be the one to do it !? My partner is already shouting! Focus

**Comment by hediet** [comment]
> It should undo in ALL files, not just open ones.

It does so. Closed files are opened automatically:

![recording](https://user-images.githubusercontent.com/2931520/138109421-4fab70d0-fbbc-4e0a-b7da-3b5c95807edb.gif)

**Comment by lukaszpolowczyk** [comment]
@RavenHursT Replace All worked for both open and unopened files.
Undo worked like "Undo Replace All" only when there is at least one of the files that was changed by Replace All open and active.

I think the context for Undo, is the active tab.

As I wrote, this is strange behavior and should be fixed.

**Comment by brunoaugusto** [comment]
This may be an indirect question to this matter but, how do I keep the files I've performed a replacement on open after the procedure? Because, currently (to me), I <kbd>Ctrl+SHIFT+H</kbd> for the action but the files only remain open if I replace each occurrence **manually**, one entry after another (thus defeating the purpose of a batch replacement). While batching, all files are closed after VSCode finished work with each one of them.

I don't know how to explain but, it's like the replacement action takes place in the background or then it's like if it happens so fast that not all files in which a replacement had been performed end up being registered in the Recently Opened Files.

Here's a hypothetical scenario to describe what just happened to me: I had like 4 files (a, b, c, d) with 3 others previously closed (e, f, g) in history. Then I needed to replace something in 5 other files (h, i, j, k, l). Action completed, affected files closed instead of remaining open as desired, but when I <kbd>Ctrl+SHIFT+T</kbd> to reopen them, the files brought to me were, let's say, `l`, `k` and `g`, instead of all five affected and only THEN the other three that were closed before the batch replacing.

Hope I didn't complicate too much with the example >.<

**Comment by HamzaAbubakar** [comment]
> > It should undo in ALL files, not just open ones.
> 
> It does so. Closed files are opened automatically:
> 
> ![recording](https://user-images.githubusercontent.com/2931520/138109421-4fab70d0-fbbc-4e0a-b7da-3b5c95807edb.gif)

This saved me from manual undo,  thanks dude :)

**Comment by lukaszpolowczyk** [comment]
> It does so. Closed files are opened automatically:

Read this comment: https://github.com/microsoft/vscode/issues/47783#issuecomment-947746697

**Comment by Olshansk** [comment]
@CodeBradley 4 more years!

**Comment by parthasarathi95** [comment]
I still get the empty array in Android while fetching products using getProducts().  NEED HELP, it's over 24 hrs I have created subscription products in play store, but still not showing

**Comment by moonjoungyoung** [comment]
Any update?
Please T.T

**Comment by LeonardoRick** [comment]
I agree, this would be a really nice feature!

**Comment by RavenHursT** [comment]
Lol.. going on 6 years later.. 

VSCode is a text editor w/ a whole slew of plugins to make it "act like" a true IDE.  Still working in JetBrains and lovin' it.

Just did a "coding exercise" on codingsandbox.io, which uses VSCode as it's basis for it's interface, and good Lawd that was painful!  I had to keep refreshing the window just to get TypeScript annotation corrections to be picked up by the editor.  Absolute garbage.. 

JetBrains is like, $10/mo to use ALL of their IDEs.. Again, #YouGetWhatYouPayFor 😉

**Comment by plcharriere** [comment]
🤡
